### PR TITLE
Don't create Instances if the nearest specialization ancestor is abstract

### LINKED ISSUE #1218: SUSHI should not allow abstract resources as profile Parents of InstanceOfs
SUSHI should not allow abstract resources, like `DomainResource` to be the `Parent` of a `Profile` or the `InstanceOf` of an `Instance`. Right now, it allows it, but then the IG Publisher emits an error.

Reported by @markkramerus on Slack.

### PR DESCRIPTION
**Pull request by mint-thompson**
Fixes #1218 and completes task [CIMPL-1070](https://standardhealthrecord.atlassian.net/browse/CIMPL-1070).

~When exporting an Instance, check if InstanceOf is abstract. If it is, log an error and do not create the Instance.~

~When exporting a Profile, check if Parent is an abstract resource. If it is, log an error and do not create the Profile. An abstract Profile is allowed as a parent, however.~

Remove unused InvalidProfileParentError.

Small updates to existing tests for InstanceExporter and StructureDefinitionExporter so that they are no longer breaking these new rules.

In the small regression set, one repo showed changes. The [fhir-sdoh-clinicalcare](https://github.com/HL7/fhir-sdoh-clinicalcare) IG contains an [Instance](https://github.com/HL7/fhir-sdoh-clinicalcare/blob/bf4880b/input/fsh/instances/SDOHCC-GroupSDOHPatientsForPayerExample.fsh) of an [abstract Profile](https://github.com/HL7/fhir-sdoh-clinicalcare/blob/bf4880b/input/fsh/profiles/SDOHCCGroup.fsh), which now throws an error.

Update 2023-03-20: Creating a Profile with an abstract parent is allowed. For Instances, look through the definition hierarchy and find the nearest "specialization" ancestor. If that definition is abstract, don't create the Instance.

### CODE REVIEW

**Review by jafeltra — APPROVED** [review]
Makes sense to me!

**Review by cmoesel — CHANGES_REQUESTED** [review]
You've implemented the fix for #1218 exactly as it was requested. Unfortunately, I probably should have thought more deeply about #1218 before submitting it.

It comes down to two things, which I elaborated on in individual comments:
* For instances, we only care about the `abstract` status of the `specialization` it is based on (not any profiles that might be in the inheritance chain).
* For profiles, it is actually completely valid to profile an abstract resource.

In doing this, I also noticed that we hard-code `abstract` to `false` when creating profiles (regardless of the parent's `abstract` value). I'm wondering if that is really the right thing to do. I don't know... What do you think?

File: src/export/InstanceExporter.ts [hunk 1]
```diff
@@ -601,6 +602,10 @@ export class InstanceExporter implements Fishable {
       }
     }
 
+    if (json.abstract === true) {
+      throw new AbstractInstanceOfError(fshDefinition.name, json.name, fshDefinition.sourceInfo);
+    }
```
[the remark this answers is not in the document]
Comment (cmoesel) [reply src/export/InstanceExporter.ts:607]: Yeah... I think that's fair. We're really just trying to help users detect an issue before they run it through the IG Publisher. So if we miss some, it's not a big deal.

**Review by cmoesel — APPROVED** [review]
You have handled the abstract art of fire well.

image

File: src/errors/InvalidProfileParentError.ts
[the remark this answers is not in the document]
Comment (jafeltra) [reply src/errors/InvalidProfileParentError.ts:1]: I love commenting on a file!

File: test/export/StructureDefinitionExporter.test.ts
Comment (jafeltra) [inline test/export/StructureDefinitionExporter.test.ts:1]: We probably don't technically need to update these tests anymore, but the changes are pretty neutral, so it is probably more work than it's worth to revert.

**Review by jafeltra — APPROVED** [review]
I like it!

**Pull request merged**